ethcore/src/verification/queue/mod.rs

```diff
@@ -420,14 +439,104 @@ impl<K: Kind> VerificationQueue<K> {
 		}
 	}
 
-	/// Optimise memory footprint of the heap fields.
+	/// Optimise memory footprint of the heap fields, and adjust the number of threads
+	/// to better suit the workload.
 	pub fn collect_garbage(&self) {
+		// number of ticks to average queue stats over
+		// when deciding whether to change the number of verifiers.
+		const READJUSTMENT_PERIOD: usize = 5;
+
 		{
 			self.verification.unverified.lock().shrink_to_fit();
 			self.verification.verifying.lock().shrink_to_fit();
 			self.verification.verified.lock().shrink_to_fit();
 		}
+
 		self.processing.write().shrink_to_fit();
+
+		if self.ticks_since_adjustment.load(AtomicOrdering::SeqCst) == READJUSTMENT_PERIOD {
+			self.ticks_since_adjustment.store(0, AtomicOrdering::SeqCst);
+		} else {
+			self.ticks_since_adjustment.fetch_add(1, AtomicOrdering::SeqCst);
+			return;
+		}
+
+		let v_count = self.verification.verified_count.load(AtomicOrdering::Acquire);
+		let drained = self.verification.drained.load(AtomicOrdering::Acquire);
+
+		self.verification.verified_count.store(0, AtomicOrdering::Release);
+		self.verification.drained.store(0, AtomicOrdering::Release);
+
+		// compute the average rate of verification per thread and determine
+		// how many are necessary to match the rate of draining.
+		let num_verifiers = self.verifiers.lock().len();
+		let v_count_per = v_count as f64 / num_verifiers as f64;
+		let needed = if v_count < 20 {
+			1
+		} else {
+			(drained as f64 / v_count_per as f64).ceil() as usize
+		};
+
+		trace!(target: "verification", "v_rate_per={}, drained={}, scaling to {} verifiers",
+			v_count_per, drained, needed);
+
+		for _ in num_verifiers..needed {
```

Comment:
can't extra 2 verifiers be added to the 7 verifiers
so that this check later will fail given ` ::num_cpus::get() == 8`
```rust
	if len == ::num_cpus::get() {
			return;
     }
```